### PR TITLE
Fixes #11684: Replace rudder-agent init script by systemd units on compatible OSes

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/11684

### CODE REVIEW

**Comment by amousset** [comment]
First version, not finished moving rudder.init to rudder-webapp.

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
The current state should be mergeable, but the rudder/rudder-server service is missing

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
@peckpeck Ready for a first review. Following PRs will include:

* the rudder/rudder-server service proper handling in rudder-webapp (not strictly necessary)
* systemd services in deb package

File: rudder-agent/SOURCES/systemd/cf-serverd.service [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[Unit]
+Description=CFEngine file server
+After=syslog.target
+After=network.target
+ConditionPathExists=/var/rudder/cfengine-community/bin/cf-serverd
+ConditionPathExists=/var/rudder/cfengine-community/inputs/promises.cf
+ConditionPathExists=!/opt/rudder/etc/disable-agent
+ReloadPropagatedFrom=rudder-agent.service
+PartOf=rudder-agent.service
```
Comment (peckpeck) [inline rudder-agent/SOURCES/systemd/cf-serverd.service]: Shouldn't this be removed to enable disabling cf-execd without disabling rudder-agent ?

File: rudder-agent/SPECS/rudder-agent.spec [hunk 2]
```diff
@@ -288,11 +291,25 @@ cd %{_sourcedir}
 %define no_ldso true
 %endif
 
+#### Use systemd everywhere except on: AIX, RHEL<7, SLES<12
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: Don't we have any other old distro like fedora ?

File: rudder-agent/SPECS/rudder-agent.spec [hunk 3]
```diff
@@ -333,24 +350,25 @@ then
   /usr/sbin/mkitab "rudder-agent:23456789:once:/usr/bin/startsrc -s rudder-agent"
   # No need to tell init to re-read /etc/inittab, it does it automatically every 60 seconds
 %else
-  RUDDER_AGENT_INIT_ENABLED=$(LANG=C chkconfig --list 2>/dev/null | grep -Ec "rudder-agent.*on")
-
-  if [ "${RUDDER_AGENT_INIT_ENABLED}" -ne 0 ]
-  then
-    chkconfig --del rudder-agent
-  fi
-  chkconfig --add rudder
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: you probably need a --del in the new prerm

Comment (peckpeck) [reply rudder-agent/SPECS/rudder-agent.spec]: and a --add rudder-agent for non systemd os

File: rudder-agent/SPECS/rudder-agent.spec [hunk 4]
```diff
@@ -333,24 +350,25 @@ then
...
-  fi
-  chkconfig rudder on
 %endif
+
   CFRUDDER_FIRST_INSTALL=1
 fi
 
+%if "%{?_os}" != "aix"
+  # Remove the rudder service
+  rm -r /etc/init.d/rudder
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: Why -r ?

File: rudder-agent/SPECS/rudder-agent.spec [hunk 5]
```diff
@@ -333,24 +350,25 @@ then
...
-  RUDDER_AGENT_INIT_ENABLED=$(LANG=C chkconfig --list 2>/dev/null | grep -Ec "rudder-agent.*on")
-
-  if [ "${RUDDER_AGENT_INIT_ENABLED}" -ne 0 ]
-  then
-    chkconfig --del rudder-agent
-  fi
-  chkconfig --add rudder
+ 
+%if "%{use_systemd}" == "true"
+  rm -r /etc/init.d/rudder-agent
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: why -r ?

File: rudder-agent/SPECS/rudder-agent.spec (hunk 2 above)
Comment (amousset) [reply rudder-agent/SPECS/rudder-agent.spec]: I will add the Fedora<15 condition.

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/SOURCES/systemd/cf-execd.service [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+Type=simple
+ExecStart=/var/rudder/cfengine-community/bin/cf-execd --no-fork
+ExecReload=/bin/kill -HUP $MAINPID
+Restart=always
+RestartSec=10
+KillMode=process
+
+[Install]
+WantedBy=multi-user.target
+WantedBy=rudder-agent.service
```
Comment (peckpeck) [inline rudder-agent/SOURCES/systemd/cf-execd.service]: Add Requiredby Rudder-agent

File: rudder-agent/SOURCES/systemd/cf-execd.service [hunk 7]
```diff
@@ -0,0 +1,20 @@
+[Unit]
```
Comment (peckpeck) [inline rudder-agent/SOURCES/systemd/cf-execd.service]: Rename to rudder-cf-execd

**Comment by amousset** [comment]
PR rebased

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
@peck Added migration scripts, ready for review

File: rudder-agent/SOURCES/systemd/rudder-agent.service [hunk 8]
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Rudder agent umbrella service
+Documentation=man:rudder(8)
+Documentation=https://www.rudder-project.org/doc
+After=syslog.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=/bin/true
```
Comment (peckpeck) [inline rudder-agent/SOURCES/systemd/rudder-agent.service]: Could comment on why  ?

File: rudder-agent/SOURCES/systemd/rudder-agent.service [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ExecReload=/bin/true
+
+[Install]
+WantedBy=multi-user.target
```
Comment (peckpeck) [inline rudder-agent/SOURCES/systemd/rudder-agent.service]: Comment on the list of dependencies

File: rudder-agent/SPECS/rudder-agent.spec [hunk 10]
```diff
@@ -304,18 +322,37 @@ find %{buildroot} -type f -o -type l | sed "s,%{buildroot},," | sed "s,\.py$,\.p
...
-  for i in init.d default cron.d; do
-    if [ -f /etc/${i}/rudder-agent ]; then
-      mkdir -p /var/backups/rudder
-      if [ "${i}" = "init.d" ]; then mode=755; else mode=644; fi
-      %{install_command} -m ${mode} /etc/${i}/rudder-agent /var/backups/rudder/rudder-agent.$(basename ${i} .d)-$(date +%Y%m%d) && echo "INFO: A back up copy of /etc/${i}/rudder-agent has been created in /var/backups/rudder"
+
+%if "%{use_systemd}" == "true"
+if [ -f /etc/init.d/rudder-agent ]
+then
+  # we are migrating from sysv to systemd
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: mkdir -p

File: rudder-agent/SPECS/rudder-agent.spec [hunk 11]
```diff
@@ -304,18 +322,37 @@ find %{buildroot} -type f -o -type l | sed "s,%{buildroot},," | sed "s,\.py$,\.p
...
+  touch /var/rudder/tmp/migration-rudder-service-upgrade
+
+  # If old rudder service is here and enabled
+  ${paths.path[grep]} -q -e 3:on -e B:on
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: remove

File: rudder-agent/SPECS/rudder-agent.spec [hunk 12]
```diff
@@ -333,24 +370,52 @@ then
...
+  /bin/systemctl daemon-reload
+
+  if [ -f /var/rudder/tmp/migration-rudder-service-enabled ]
   then
-    chkconfig rudder-agent off
+    /bin/systemctl enable rudder-agent cf-execd
+
+    if [ -f /var/rudder/tmp/migration-rudder-cf-serverd-disabled ]
+    then
+      # Keep cf-serverd disabled
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: /bin/true

File: rudder-agent/SPECS/rudder-agent.spec [hunk 13]
```diff
@@ -304,18 +322,37 @@ find %{buildroot} -type f -o -type l | sed "s,%{buildroot},," | sed "s,\.py$,\.p
...
 if [ $1 -eq 2 ];then
-%if "%{?_os}" != "aix"
-  # Keep a backup copy of Rudder agent init and cron files to prevent http://www.rudder-project.org/redmine/issues/3995
-  for i in init.d default cron.d; do
-    if [ -f /etc/${i}/rudder-agent ]; then
-      mkdir -p /var/backups/rudder
-      if [ "${i}" = "init.d" ]; then mode=755; else mode=644; fi
-      %{install_command} -m ${mode} /etc/${i}/rudder-agent /var/backups/rudder/rudder-agent.$(basename ${i} .d)-$(date +%Y%m%d) && echo "INFO: A back up copy of /etc/${i}/rudder-agent has been created in /var/backups/rudder"
+
+%if "%{use_systemd}" == "true"
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: You should do this even without systemd

File: rudder-agent/SPECS/rudder-agent.spec [hunk 14]
```diff
@@ -333,24 +370,52 @@ then
...
+    fi
   fi
-  chkconfig rudder on
-%endif
-  CFRUDDER_FIRST_INSTALL=1
+  
+  rm -f /var/rudder/tmp/migration-rudder-service-upgrade
+  rm -f /var/rudder/tmp/migration-rudder-service-enabled
+  rm -f /var/rudder/tmp/migration-rudder-cf-serverd-disabled
 fi
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: Reenable rudder-agent on system without systemd

File: rudder-agent/debian/rules [hunk 15]
```diff
@@ -83,7 +82,7 @@ binary-arch: install
 #	dh_installmime
 #	dh_python
 	# Init script and configuration files
-	dh_installinit --name=rudder --no-start -R # restart after upgrade (instead of stop, upgrade, start)
+	dh_installinit --name=rudder-agent --no-start -R # restart after upgrade (instead of stop, upgrade, start)
```
Comment (peckpeck) [inline rudder-agent/debian/rules]: name is now the default

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/SOURCES/systemd/rudder-agent.service (hunk 9 above)
Comment (peckpeck) [reply rudder-agent/SOURCES/systemd/rudder-agent.service]: Add comments on the list of dependencies

File: rudder-agent/SPECS/rudder-agent.spec [hunk 16]
```diff
@@ -304,17 +322,39 @@ find %{buildroot} -type f -o -type l | sed "s,%{buildroot},," | sed "s,\.py$,\.p
...
-      if [ "${i}" = "init.d" ]; then mode=755; else mode=644; fi
-      %{install_command} -m ${mode} /etc/${i}/rudder-agent /var/backups/rudder/rudder-agent.$(basename ${i} .d)-$(date +%Y%m%d) && echo "INFO: A back up copy of /etc/${i}/rudder-agent has been created in /var/backups/rudder"
+
+# Part of the package now, but create it anyway in case
+# of upgrade from a very old version
+mkdir -p /var/rudder/tmp
+
+if [ -f /etc/init.d/rudder ] || [ -f /etc/init.d/rudder-agent ]
+then
+  # We are migrating from a pre-4.3 to 4.3
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: Or to a systemd-less OS

File: rudder-agent/SPECS/rudder-agent.spec [hunk 17]
```diff
@@ -304,17 +322,39 @@ find %{buildroot} -type f -o -type l | sed "s,%{buildroot},," | sed "s,\.py$,\.p
...
+
+  # If old rudder service is here and enabled
+  if chkconfig --list rudder 2>&1 | grep -q -e 3:on -e B:on
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: chkconfig doesn't exist on aix

File: rudder-agent/SPECS/rudder-agent.spec [hunk 18]
```diff
@@ -333,24 +373,58 @@ then
...
 %else
-  RUDDER_AGENT_INIT_ENABLED=$(LANG=C chkconfig --list 2>/dev/null | grep -Ec "rudder-agent.*on")
+  %if "%{use_systemd}" == "true"
+    /bin/systemctl daemon-reload
+    /bin/systemctl enable rudder-agent cf-execd cf-serverd
+  %else
+    chkconfig --add rudder-agent
+    chkconfig rudder-agent on
+  %endif
+%endif
```
Comment (peckpeck) [inline rudder-agent/SPECS/rudder-agent.spec]: Isn't this %endif redudant ?

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Finally I will make the debian change too.

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 19]
```diff
@@ -2,20 +2,97 @@
 set -e
 
 CFRUDDER_FIRST_INSTALL="$1"
+CFRUDDER_USE_SYSTEMD="$2"
+CFRUDDER_AIX="$3"
+CFRUDDER_DEB="$4"
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: acan't you merge both into a CFRUDDER_OS ? value debian/redhat/aix

File: rudder-agent/SOURCES/rudder-agent-preinst [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+echo "`date` - Starting rudder-agent pre installation script" >> ${LOG_FILE}
+
+# Restart daemons if we stopped them, otherwise not
+if [ ${CFRUDDER_FIRST_INSTALL} -ne 1 ]
+then
+  # Part of the package now, but create it anyway in case
+  # of upgrade from a very old version
+  mkdir -p /var/rudder/tmp
+
+  %if "%{use_systemd}" == "true"
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-preinst]: ?

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 21]
```diff
@@ -2,20 +2,114 @@
...
+# Do this at first install
+if [ ${CFRUDDER_FIRST_INSTALL} -eq 1 ]
+then
+  # Set rudder-agent as service
+  if [ ${CFRUDDER_OS} = "aix" ]
+  then
+    /usr/bin/mkssys -s rudder-agent -p %{ruddervardir}/cfengine-community/bin/cf-execd -a "-F" -u root -S -n15 -f9 -R
+    /usr/sbin/mkitab "rudder-agent:23456789:once:/usr/bin/startsrc -s rudder-agent"
+    # No need to tell init to re-read /etc/inittab, it does it automatically every 60 seconds
+  else
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: an elif would be more clear here

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 22]
```diff
@@ -2,20 +2,114 @@
...
+      else
+        chkconfig --add rudder-agent
+        chkconfig rudder-agent on
+      fi
+    fi
+  fi
+fi
+
+if [ -f /var/rudder/tmp/migration-rudder-service-rename ]
+then
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: Add a comment to say we are not in the systemd case here

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 23]
```diff
@@ -2,20 +2,114 @@
...
+    then
+      chkconfig --add rudder-agent
+      chkconfig rudder-agent on
+    fi
+
+    if type update-rc.d > /dev/null
+    then
+      update-rc.d rudder-agent enable
+    fi
+  fi
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: What about the rename but not enabled case ?

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 24]
```diff
@@ -2,20 +2,114 @@
...
+then
+  # We are making the migration from sysv to systemd
+  if type chkconfig > /dev/null
+  then
+    chkconfig --del rudder-agent
+  fi
+
+  if type update-rc.d > /dev/null
+  then
+    update-rc.d rudder-agent disable
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: i think it should be remove

File: rudder-agent/SOURCES/rudder-agent-postinst [hunk 25]
```diff
@@ -2,20 +2,114 @@
...
+# Remove the rudder service
+if [ -f /etc/init.d/rudder ]
+then
+  if type chkconfig > /dev/null
+  then
+    chkconfig --del rudder || true
+  fi
+  if type update-rc.d > /dev/null
+  then
+    update-rc.d rudder disable
```
Comment (peckpeck) [inline rudder-agent/SOURCES/rudder-agent-postinst]: "remove" here too

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/debian/preinst [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+case "$1" in
+    install)
+        CFRUDDER_FIRST_INSTALL=1
+    ;;
+
+    upgrade)
+
+    ;;
+
+    abort-upgrade)
```
[the remark this answers is not in the document]
Comment (peckpeck) [reply rudder-agent/debian/preinst]: an exit is mandatory here

**Comment by amousset** [comment]
Commit modified

File: rudder-agent/debian/preinst [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+
+if type systemctl > /dev/null ; then CFRUDDER_USE_SYSTEMD=1; fi
+
+case "$1" in
+    install)
+        CFRUDDER_FIRST_INSTALL=1
+    ;;
+
+    upgrade)
+	exit 0
```
Comment (peckpeck) [inline rudder-agent/debian/preinst]: No exit here

**Comment by amousset** [comment]
Commit modified

**Pull request merged**